### DISCUSSION TITLE
encounter memory leak with uvicorn worker

### DISCUSSION THREAD

**Discussion by kevin1997998** [fastapi/fastapi#11992] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import json
import os
import uvicorn
from fastapi import FastAPI

app = FastAPI()

@app.post("/get-all-order-item")
def get_all_order_item():
    file_path = os.path.join("./mock", f"get-all-order-item.json")
    with open(file_path, "r") as file:
        mock_response = json.load(file)
    return mock_response

if __name__ == "__main__":
    uvicorn.run("app:app", host="0.0.0.0", port=5000, reload=False, workers=8)
```

### Description

above is my simple experiment with memory leak
get-all-order-item.json is a json file with about 30MB
i don't even use locust, just periodically makes api calls with the script below (around 300 ~ 400 calls)
I already observe memory leak in local and even deployed on k8s
```
import time
import requests
import os
print(os.getpid())
while 1:
    response = requests.post("http://localhost:5000/get-all-order-item", json={}) 
    time.sleep(2)
```
did anyone also encounter memory leak problem using fastapi?

### Operating System

macOS

### FastAPI Version

0.110.1

### Pydantic Version

2.8.2

### Python Version

3.12.4

**Comment by iamdhakrey** [comment]
The code you've provided is a simple FastAPI application that loads a 30MB JSON file on each request to the `/get-all-order-item` endpoint. Based on your description, you're observing a memory leak after making around 300-400 API calls, both locally and on Kubernetes (K8s).

### Potential Causes of the Memory Leak:
1. **File Handling**: The memory leak might be due to how the file is being opened and read. If there's any issue in how the file is managed (like not being closed properly, though `with open()` should handle that), it might cause a memory leak over time.

2. **Large File Loading**: Loading a large file repeatedly into memory (30MB each time) could cause memory usage to increase, especially if the garbage collector doesn’t free up memory quickly enough between requests.

3. **Concurrency and Workers**: Since you're running with 8 workers, each worker might be holding onto resources, leading to cumulative memory usage.

### Suggestions to Mitigate Memory Leak:
1. **File Caching**: Consider loading the file once during startup and reusing the loaded data across all requests.

    ```python
    import json
    import os
    import uvicorn
    from fastapi import FastAPI

    app = FastAPI()

    # Load the JSON file once during startup
    file_path = os.path.join("./mock", "get-all-order-item.json")
    with open(file_path, "r") as file:
        mock_response = json.load(file)

    @app.post("/get-all-order-item")
    def get_all_order_item():
        return mock_response

    if __name__ == "__main__":
        uvicorn.run("app:app", host="0.0.0.0", port=5000, reload=False, workers=8)
    ```

    This way, the file is loaded into memory once, and each request returns the same in-memory data.

2. **Memory Profiling**: Use memory profiling tools like `memory_profiler` to see exactly where memory usage is increasing. This can give more insight into where the problem lies.

    ```bash
    pip install memory_profiler
    ```

    Then you can use `@profile` decorators or `mprof` to track memory usage.

3. **Garbage Collection**: Force garbage collection after each request (this is more of a temporary workaround than a solution).

    ```python
    import gc
    
    @app.post("/get-all-order-item")
    def get_all_order_item():
        with open(file_path, "r") as file:
            mock_response = json.load(file)
        gc.collect()
        return mock_response
    ```

4. **Concurrency**: Reduce the number of workers or test with different worker settings to see if it impacts memory usage.

**Comment by YuriiMotov** [comment]
I tried running your tests and I can NOT reproduce memory leak..
It was probably fixed somewhere.

Could you check whether the problem still persists for you?

I'll close this for now to keep the discussions clean.
If you still need help, feel free to re-open the discussion.
